### PR TITLE
Cards test

### PR DESCRIPTION
**Pull request by istzen**
Tests for Cards, note that the coverage is not that good since we have no method to test the color if I am correct, otherwise please let me know then I'll add tests to check for the status color.

Add pumpScaffold method for all widgets that don't have a scaffold themselves.

### CODE REVIEW

File: lib/util/trip/trip_card.dart [hunk 1]
```diff
@@ -180,9 +182,7 @@ abstract class TripCardState<T extends Trip, U extends TripCard<T>> extends Stat
     return const SizedBox();
   }
 
-  Widget buildRightSide() {
-    return const SizedBox();
-  }
+  Widget buildRightSide();
```
Comment (fynsta) [inline lib/util/trip/trip_card.dart:185]: wait why did you change this? some other card might not want to implement this

Comment (istzen) [reply lib/util/trip/trip_card.dart:185]: Because it's overwritten in every subclass but i can change it Back.

Comment (fynsta) [reply lib/util/trip/trip_card.dart:185]: Yeah but that's not guaranteed right?

File: test/widgets/cards_test.dart [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+  setUp(() {
+    reset(processor);
+  });
+
+  //since the code is the same for RideCard DriveCard and PendingRideCard, we only test DriveCard
+  testWidgets('TripCard shows the correct information', (WidgetTester tester) async {
+    drive = DriveFactory().generateFake();
+    whenRequest(processor).thenReturnJson(drive.toJsonForApi());
+    await pumpMaterial(tester, DriveCard(drive, key: const Key('driveCard')));
```
Comment (fynsta) [inline test/widgets/cards_test.dart:45]: what's the point of passing that key?

File: test/widgets/cards_test.dart [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+      final Finder profileWidgetFinder = find.byType(ProfileWidget);
+      expect(profileWidgetFinder, findsOneWidget);
+      final ProfileWidget profileWidget = tester.widget(profileWidgetFinder);
+      expect(profileWidget.profile, ride.rider);
+
+      //Extra time is shown
+      expect(find.byKey(const Key('extraTime')), findsOneWidget);
+
+      //Seats indicator is shown
+      final Finder seatsIndicator = find.byType(IconWidget);
```
Comment (fynsta) [inline test/widgets/cards_test.dart:88]: Maybe better to check for an actual `SeatsIndicator` object.

File: test/widgets/cards_test.dart [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+      expect(find.byKey(const Key('approveButton')), findsOneWidget);
+
+      //Reject button is shown
+      expect(find.byKey(const Key('rejectButton')), findsOneWidget);
+    });
+
+    group('approve Ride', () {
+      Future<void> openApproveDialog(WidgetTester tester) async {
+        //need scaffold for dialog
+        await pumpMaterial(tester, Scaffold(body: PendingRideCard(ride, reloadPage: () {}, drive: drive)));
```
Comment (fynsta) [inline test/widgets/cards_test.dart:105]: tbh maybe we should just add a Scaffold by default to `pumpMaterial`

File: test/widgets/cards_test.dart [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        await tester.tap(approveButton);
+        //load dialog
+        await tester.pumpAndSettle();
+      }
+
+      testWidgets('can approve Ride if possible', (WidgetTester tester) async {
+        await openApproveDialog(tester);
+
+        //confirm dialog
+        final Finder confimrButton = find.byKey(const Key('approveConfirmButton'));
```
Comment (fynsta) [inline test/widgets/cards_test.dart:117]: typo

File: lib/util/trip/pending_ride_card.dart [hunk 6]
```diff
@@ -126,6 +130,7 @@ class _PendingRideCardState extends TripCardState<Ride, PendingRideCard> {
                 Navigator.of(dialogContext).pop();
                 ScaffoldMessenger.of(context).showSnackBar(
                   SnackBar(
+                    key: const Key('approveSuccesSnackbar'),
```
Comment (fynsta) [inline lib/util/trip/pending_ride_card.dart:133]: success is spelled with 2x "s"

File: test/widgets/cards_test.dart [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+          ProfileFeatureFactory().generateFake(rank: 1, feature: Feature.accessible),
+          ProfileFeatureFactory().generateFake(rank: 2, feature: Feature.noSmoking),
+          ProfileFeatureFactory().generateFake(rank: 3, feature: Feature.noVaping),
+          ProfileFeatureFactory().generateFake(rank: 4, feature: Feature.noPetsAllowed),
+        ]);
+
+        expect(find.byIcon(Icons.accessibility), findsOneWidget);
+        expect(find.byIcon(Icons.smoke_free), findsOneWidget);
+        expect(find.byIcon(Icons.vape_free), findsOneWidget);
+
```
Comment (fynsta) [inline test/widgets/cards_test.dart:269]: there should be a test that there is no `pets` icon

File: test/widgets/cards_test.dart [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        ///7 Icons are always shown
+        expect(find.byType(Icon), findsNWidgets(10));
```
Comment (fynsta) [inline test/widgets/cards_test.dart:271]: I would remove this, that line would always have to be changed when we change completely unrelated parts of the card. Probably better to give the feature row a key and check that there are 3 icon descendants

File: test/widgets/cards_test.dart [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+        await tester.pumpAndSettle();
+
+        expect(find.byKey(const Key('rejectSuccessSnackbar')), findsNothing);
+        verifyRequestNever(processor, urlMatcher: equals('/rest/v1/rpc/reject_ride'));
+      });
+    });
+  });
+
+  group('RideCard', () {
+    Future<void> loadRideCard(WidgetTester tester, List<ProfileFeature>? features) async {
```
Comment (fynsta) [inline test/widgets/cards_test.dart:225]: features should probably not be a positional parameter, it should be named

**Comment by fynsta** [comment]
Regarding testing color. I just looked up how you do that.

First, you need to get BuildContext of the app. That can be done with
`final BuildContext context = tester.element(find.byType(Container).first);`

Then you find whatever element you want to get the color of. For example
`final Finder accessibleIcon = find.byIcon(Icons.accessibility);`

And finally you query the color attribute of the corresponding widget:
`expect(tester.widget<Icon>(accessibleIcon).color, Theme.of(context).colorScheme.primary);`

File: lib/util/trip/trip_card.dart (hunk 1 above)
Comment (Eddie-42) [reply lib/util/trip/trip_card.dart:185]: I think the change is fine. It serves as a reminder that TripCard is, indeed, abstract, and if an extending class doesn't want anything there, it should implement that itself. This just forces us to think about it

File: lib/util/trip/trip_card.dart [hunk 10]
```diff
@@ -180,9 +182,7 @@ abstract class TripCardState<T extends Trip, U extends TripCard<T>> extends Stat
     return const SizedBox();
```
Comment (Eddie-42) [inline lib/util/trip/trip_card.dart:182]: IMO, we should do the same thing as buildRightSide here then

File: test/widgets/cards_test.dart [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+          methodMatcher: equals('POST'),
+          bodyMatcher: equals({'ride_id': ride.id}),
+        ).called(1);
+      });
+
+      testWidgets('can not approve Ride if not possible', (WidgetTester tester) async {
+        drive = DriveFactory().generateFake(rides: [ride], seats: 1);
+        ride = RideFactory().generateFake(
+          status: RideStatus.pending,
+          rider: NullableParameter(ProfileFactory().generateFake(id: 1)),
```
Comment (Eddie-42) [inline test/widgets/cards_test.dart:137]: Why is the ID needed?

File: test/widgets/cards_test.dart [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        expect(find.byKey(const Key('approveSuccesSnackbar')), findsOneWidget);
+        verifyRequest(
+          processor,
+          urlMatcher: equals('/rest/v1/rpc/approve_ride'),
+          methodMatcher: equals('POST'),
+          bodyMatcher: equals({'ride_id': ride.id}),
+        ).called(1);
+      });
+
+      testWidgets('can not approve Ride if not possible', (WidgetTester tester) async {
```
Comment (Eddie-42) [inline test/widgets/cards_test.dart:133]: Maybe include something along the lines of "too many seats" instead of "not possible" for more clarity when just reading the error messages

File: test/widgets/cards_test.dart (hunk 5 above)
Comment (Eddie-42) [reply test/widgets/cards_test.dart:117]: In the next test, too

File: lib/util/trip/trip_card.dart (hunk 1 above)
Comment (fynsta) [reply lib/util/trip/trip_card.dart:185]: No? It's much better if you just fill the spaces that you want. The rest should obviously always be empty boxes, so that should also be the default given in `TripCard`.

Comment (Eddie-42) [reply lib/util/trip/trip_card.dart:185]: I see two different philosophies here. I like it explicit, you like it implicit. That's fine, but I like it explicit

Comment (istzen) [reply lib/util/trip/trip_card.dart:185]: okay so what do i do now? should i change it back or not? what do the others think @Showcas @MarkFalke

File: test/widgets/cards_test.dart (hunk 2 above)
Comment (istzen) [reply test/widgets/cards_test.dart:45]: good question xD

File: test/widgets/cards_test.dart (hunk 3 above)
Comment (istzen) [reply test/widgets/cards_test.dart:88]: yes good idea thansk

Comment (istzen) [reply test/widgets/cards_test.dart:88]: actually i just looked into the code, there is nor `SeatIndicator` object there, but a `IconWidget` with icons.chair that's why i test it that way. I can however change it to a `SeatIndicator` not sure if that's the best way. 
@Showcas can you explain why there is a `IconWidget` used here and no `SeatIndicator`?

File: test/widgets/cards_test.dart (hunk 4 above)
Comment (istzen) [reply test/widgets/cards_test.dart:105]: not sure since for the pages there already is a scaffold i think then it would be better to just add either a parameter `hasScaffold` or something or just use two methods

File: test/widgets/cards_test.dart (hunk 5 above)
Comment (istzen) [reply test/widgets/cards_test.dart:117]: thanks

File: lib/util/trip/pending_ride_card.dart (hunk 6 above)
Comment (istzen) [reply lib/util/trip/pending_ride_card.dart:133]: thanks

File: test/widgets/cards_test.dart (hunk 7 above)
Comment (istzen) [reply test/widgets/cards_test.dart:269]: thanks i had that not sure why i removed it xD

File: test/widgets/cards_test.dart (hunk 9 above)
Comment (istzen) [reply test/widgets/cards_test.dart:225]: i don't understad what you mean exactly

File: test/widgets/cards_test.dart (hunk 12 above)
Comment (istzen) [reply test/widgets/cards_test.dart:133]: did that

File: test/widgets/cards_test.dart (hunk 11 above)
Comment (istzen) [reply test/widgets/cards_test.dart]: which id? xD

File: test/widgets/cards_test.dart (hunk 8 above)
Comment (istzen) [reply test/widgets/cards_test.dart]: yes that's a very good idea thanks

File: test/widgets/cards_test.dart (hunk 9 above)
Comment (istzen) [reply test/widgets/cards_test.dart]: i added some more xD

**Comment by istzen** [comment]
@fynsta thank you very much for the help with the colors i added test for them also so the coverage is over 90% now for each card. (only for the did updateWidget method it's missing).

File: test/widgets/cards_test.dart (hunk 4 above)
Comment (fynsta) [reply test/widgets/cards_test.dart:106]: Yes but when we write Widget tests that are not pages, we will always need this

File: test/widgets/cards_test.dart (hunk 3 above)
Comment (fynsta) [reply test/widgets/cards_test.dart:89]: I think Simon and I forgot to change it there when we added `SeatIndicator`. Just change it as part of this PR, that should have been like that in the first place.

File: test/widgets/cards_test.dart (hunk 4 above)
Comment (fynsta) [reply test/widgets/cards_test.dart:106]: Or a `pumpScaffold` helper?

File: lib/util/trip/trip_card.dart (hunk 1 above)
Comment (fynsta) [reply lib/util/trip/trip_card.dart:185]: In any case, you need to do something. Either you remove all of these "build<Position>" methods and add the stubs to the descendants. Or you just remove this change and implement it in this class again.

File: test/widgets/cards_test.dart (hunk 4 above)
Comment (istzen) [reply test/widgets/cards_test.dart:106]: yes that's what i meant. should i add it in this PR?

File: lib/util/trip/trip_card.dart (hunk 1 above)
Comment (istzen) [reply lib/util/trip/trip_card.dart:185]: ok i guess i just change it back then xD

File: test/widgets/cards_test.dart (hunk 3 above)
Comment (istzen) [reply test/widgets/cards_test.dart:89]: yes but they do different things. `IconWidget` shows if the count is bigger than two only one icon and the count and `SeatIndicator` shows just the number of icons and i think the first one is better for a card

File: test/widgets/cards_test.dart (hunk 4 above)
Comment (fynsta) [reply test/widgets/cards_test.dart:106]: Let's do it :D

File: test/widgets/cards_test.dart (hunk 3 above)
Comment (fynsta) [reply test/widgets/cards_test.dart:89]: Ah I see. In this case maybe we should still do this via the SeatIndicator with an optional boolean parameter (and that can just call IconWidget when it is true). But should be done in later refactorings of course.

**Review by fynsta — APPROVED** [review]

**Review by Eddie-42 — APPROVED** [review]

**Review by MarkFalke — APPROVED** [review]

**Review by Showcas — APPROVED** [review]
